PORT-9395 | Update sonarqube guide to use search relation (#1442)
port-labs/port-docs: docs/build-your-software-catalog/sync-data-to-catalog/code-quality-security/sonarqube/guides/connect-github-pr-with-sonar-analysis.md

<br/><br/>

Now that the <PortTooltip id="blueprint">blueprints</PortTooltip> are related,
we need to assign the relevant SonarQube analysis to each of our pull requests.
This can be done by adding some mapping logic.
Go to your [data sources page](https://app.getport.io/settings/data-sources),
and click on your GitHub Pull Request integration:
<img src='/img/guides/githubPrIntegrationDataSource.png' border='1px' />

<br/><br/>

> # Relation Mapping

Reviewer: got it !

  <img src='/img/guides/githubExporter.png' width='100%' border='1px' />
<br></br>

- Update the mapping YAML with this content:

  ```yaml showLineNumbers
  resources:
  - kind: pull-request
    selector:
    query: 'true'
    port:
    entity:
    mappings:
    identifier: .head.repo.name + '-' + (.number|tostring)
    title: .title
    blueprint: '"githubPullRequest"'
    properties:
    creator: .user.login
    assignees: '[.assignees[].login]'
    reviewers: '[.requested_reviewers[].login]'
    status: .status
    closedAt: .closed_at
    updatedAt: .updated_at
    mergedAt: .merged_at
    prNumber: .id
    link: .html_url
    relations:
    sonarAnalysis:
    combinator: '"and"'
    rules:
    - property: '"$title"'
      operator: '"="'
      value: .title
    - property: '"branch"'
      operator: '"="'
      value: .head.ref

  ```

This configuration uses the `title` and `branch` properties to establish a relationship with SonarQube analysis based on matching properties.

## Matching Pull Request Commit SHA

To establish the relationship using the `commitSha` property:
   
- Click on Builder on the top right conner

> - Go to the `SonarQube` <PortTooltip id="blueprint">blueprints</PortTooltip>.

Reviewer 2: ```suggestion
- Go to the `SonarQube` <PortTooltip id="blueprint">blueprint</PortTooltip>.
```

> - Hover over it, click on the `...` button on the right, and select `Edit`.

Reviewer 2: We need to be clearer in the instructions here. If we want to instruct the user to edit the JSON, there is no need to click edit, and then edit JSON. The guide should instruct to click on the `...` and click edit JSON. This will save the user unnecessary steps.
Also, it's unclear from the instructions if the user needs to add the mapping or replace the existing mapping. Please clarify.

Author: Done

   
<br></br>
- Update the mapping YAML with this content

```yaml showLineNumbers
  resources:
     - kind: pull-request
       selector:
          query: 'true'
       port:
          entity:
             mappings:
                identifier: .head.repo.name + '-' + (.number|tostring)
                title: .title
                blueprint: '"githubPullRequest"'
                properties:
                   creator: .user.login
                   assignees: '[.assignees[].login]'
                   reviewers: '[.requested_reviewers[].login]'
                   status: .status
                   closedAt: .closed_at
                   updatedAt: .updated_at
                   mergedAt: .merged_at
                   prNumber: .id
                   link: .html_url
                relations:
                   sonarAnalysis:
                      combinator: '"and"'
                      rules:
                         - property: '"commitSha"'
                           operator: '"="'
                           value: .head.sha

```

:::tip Mapping explanation

In this implementation,
we used search relation features

> to map relationships between `githubPullRequest` entities and `sonarAnalysis` entities based on specific **properties** rather than direct identifiers.

Reviewer 2: ```suggestion
to map relationships between `githubPullRequest` entities and `sonarAnalysis` entities based on specific **properties** rather than direct identifiers.  
```

> By using **properties** like `title`, `branch`, and `commitSha`, we are able to establish a between the two entities.

Reviewer 2: ```suggestion
By using **properties** like `title`, `branch`, and `commitSha`, we are able to establish a between the two entities.  
```

<br/><br/>

Now that the <PortTooltip id="blueprint">blueprints</PortTooltip> are related,
we need to assign the relevant SonarQube analysis to each of our pull requests.
This can be done by adding some mapping logic.
Go to your [data sources page](https://app.getport.io/settings/data-sources),
and click on your GitHub Pull Request integration:
<img src='/img/guides/githubPrIntegrationDataSource.png' border='1px' />

<br/><br/>

## Relation Mapping

> There are two ways to establish the relationship between `pull-request` kind and the `sonarqube-analysis` kind:

Reviewer 3: ```suggestion
There are two ways to establish the relationship between a GitHub `pull-request` entity and a SonarQube `analysis` entity:
```

Author: done

> <img src='/img/guides/pullReqSonarQubeRelationship.png' width='70%' border='1px' />

Reviewer 3: It is not clear what you're trying to convey with this image. Maybe we don't need it?
![image](https://github.com/user-attachments/assets/c4acf4cf-f6e0-414b-ae90-90697ce7dc74)

<br/><br/>

> 1. **Option 1** - using the `title` and `branch` **property** to perform a search query[ rule](/search-and-query/#rules) to map relations based on the **property** of the related entity:

Reviewer 3: Instead of organizing this as a list, I want it to be a sub header
### Matching common pull request title and branch

### Matching pull request commit sha

> 1. **Option 1** - using the `title` and `branch` **property** to perform a search query[ rule](/search-and-query/#rules) to map relations based on the **property** of the related entity:

Reviewer 3: Also, there is not much information on the why. You are talking about the how. Users need to understand why we are using the title and branch.

   - Go to your [data sources page](https://app.getport.io/settings/data-sources) and click on the GitHub exporter:

>       <img src='/img/githubExporter.png' width='100%' border='1px' />

Reviewer 3: This image is not rendering properly
<img width="892" alt="Screenshot 2024-07-31 at 8 44 01 AM" src="https://github.com/user-attachments/assets/6a9ce710-52bf-4a3e-ba1c-ddbc0947ca5c">

   - Add the following entry to the mapping YAML:

     ```yaml showLineNumbers
     relations:
     sonarAnalysis:
        combinator: '"and"'
        rules:
           - property: '"$title"'
             operator: '"="'
             value: .title

>            - property: '"$branch"'

Reviewer 3: $branch will not work. just use '"branch"' since it is not a meta property. [See examples of meta properties](https://docs.getport.io/build-your-software-catalog/customize-integrations/configure-data-model/setup-blueprint/properties/meta-properties). These are the only one you can use the $ to access them

             operator: '"="'
             value: .head.ref

>      ```

Reviewer 3: instead of just showing a truncated config mapping, provide the full mapping for the pull-request kind including the search rules. Then ask them to update their pull-request kind with this new one

     Now this uses the `title` and `branch` as the parameters for the search rule
     to establish a relationship with sonarQube analysis based on any of it's properties
     matching these and having equal value.

> 2. **Option 2** - modify the `sonarQube Analysis` <PortTooltip id="blueprint">blueprints</PortTooltip> to add a `commitSha` **property**:

Reviewer 3: Use the sub header I mentioned above instead of a list. and add a line text explaining what and why we are using the commit sha

   
   - Click on Builder on the top right conner
   - Go to the `SonarQube` <PortTooltip id="blueprint">blueprints</PortTooltip>.
   - Hover over it, click on the `...` button on the right, and select `Edit`.
<br></br>
   <img src='/img/guides/sonarQubeAnalysisBlueprintUpdate.png' width='40%' border='1px' />

   - Click on `Edit JSON`, add the `commitSha` **property** and `Save`:
   
     <img src='/img/guides/editSonarQubeAnalysisJson.png' width='40%' border='1px' />
   
<br></br>

<details>
<summary>Relation mapping (click to expand)</summary>

```yaml showLineNumbers
  relations:
     sonarAnalysis:
        combinator: '"and"'
        rules:

>            - property: '"$commitSha"'

Reviewer 3: ```suggestion
           - property: '"commitSha"'
```

             operator: '"="'
             value: .head.sha

> ```

Reviewer 3: Provide the full config mapping
